src/main/scala/mesosphere/marathon/core/task/Task.scala

```diff
@@ -194,98 +305,77 @@ object Task {
-        case ResidentTaskIdRegex(runSpecId, _, uuid, _, attempt) => Some(attempt.toLong)
-        case _ => None
-      }
-    }
-
-    def reservationId(taskId: String): String = {
-      taskId match {
-        case ResidentTaskIdWithInstanceIdRegex(runSpecId, prefix, uuid, container, attempt) =>
-          Reservation.Id(runSpecId, ".", Some(prefix), uuid).toString
-        case TaskIdWithInstanceIdRegex(runSpecId, prefix, instanceUuid, uuid) =>
-          Reservation.Id(runSpecId, ".", Some(prefix), instanceUuid).toString
-        case ResidentTaskIdRegex(runSpecId, separator, uuid, _, attempt) =>
-          Reservation.Id(runSpecId, separator, None, uuid).toString
-        case LegacyTaskIdRegex(runSpecId, uuid) =>
-          Reservation.Id(runSpecId, ".", None, uuid).toString
-        case _ => taskId
-      }
-    }
-
-    def instanceId(taskId: String): Instance.Id = {
-      taskId match {
-        case ResidentTaskIdWithInstanceIdRegex(runSpecId, prefix, uuid, container, attempt) =>
-          Instance.Id(runSpecId + "." + prefix + uuid)
-        case TaskIdWithInstanceIdRegex(runSpecId, prefix, instanceUuid, uuid) =>
-          Instance.Id(runSpecId + "." + prefix + instanceUuid)
-        case ResidentTaskIdRegex(runSpecId, _, uuid, _, attempt) =>
-          Instance.Id(runSpecId + "." + calculateLegacyExecutorId(uuid))
-        case LegacyTaskIdRegex(runSpecId, uuid) =>
-          Instance.Id(runSpecId + "." + calculateLegacyExecutorId(uuid))
-        case _ => throw new MatchError(s"taskId $taskId is no valid identifier")
+    /**
+      * Parse instance and task id from idString.
+      *
+      * The string has to match one of
+      *   * LegacyTaskIdRegex
+      *   * ResidentTaskIdRegex
+      *   * TaskIdWithInstanceIdRegex
+      *   * ResidentTaskIdWithInstanceIdRegex
+      *
+      * @param idString The raw id that should be parsed.
+      * @return Task.Id
+      */
+    def fromIdString(idString: String): Task.Id = {
+      idString match {
+        case ResidentTaskIdWithInstanceIdRegex(safeRunSpecId, prefix, uuid, container, attempt) =>
+          val runSpec = PathId.fromSafePath(safeRunSpecId)
+          val instanceId = Instance.Id(runSpec, Instance.Prefix.fromString(prefix), UUID.fromString(uuid))
+          val containerName: Option[String] = if (container == Names.anonymousContainer) None else Some(container)
+          ResidentTaskId(instanceId, containerName, attempt.toLong)
+        case TaskIdWithInstanceIdRegex(safeRunSpecId, prefix, uuid, container) =>
+          val runSpec = PathId.fromSafePath(safeRunSpecId)
+          val instanceId = Instance.Id(runSpec, Instance.Prefix.fromString(prefix), UUID.fromString(uuid))
+          val containerName: Option[String] = if (container == Names.anonymousContainer) None else Some(container)
+
+          // This could also be a ResidentTaskId
+          EphemeralOrReservedTaskId(instanceId, containerName)
```

Comment:
Unfortunately it seems that a reservation id has the same form as the ephemeral task id. So we cannot distinguish.

Reply:
I confirm.